tests/various/bug1531 fails when building with Fedora build flags

## Steps to reproduce the issue
I'm maintaining yosys in the Fedora RPM repos, which uses a specific set of flags to isolate potential problems and to "fortify" resulting binaries during build. Here's how I build yosys:
```
git clone https://github.com/YosysHQ/yosys
...
cd yosys
git checkout 07a12ebd
make config-gcc
export CFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection'
export CXXFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection'
export FFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -I/usr/lib64/gfortran/modules'
export FCFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -I/usr/lib64/gfortran/modules'
export LDFLAGS='-Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld'
make -j PREFIX=/usr PRETTY=0 all
...
make test SEED=314159265359
...
```

## Expected behavior

All tests pass.

## Actual behavior

```
...
Running bug1531.ys..
/usr/include/c++/9/bits/stl_vector.h:1042: std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](std::vector<_Tp, _Alloc>::size_type) [with _Tp = Yosys::RTLIL::State; _Alloc = std::allocator<Yosys::RTLIL::State>; std::vector<_Tp, _Alloc>::reference = Yosys::RTLIL::State&; std::vector<_Tp, _Alloc>::size_type = long unsigned int]: Assertion '__builtin_expect(__n < this->size(), true)' failed.
make[1]: *** [run-test.mk:29: run-bug1531.ys] Aborted (core dumped)
make[1]: Leaving directory '/home/somlo/yosys/tests/various'
make: *** [Makefile:716: test] Error 2
```

Bisecting the problem leads me to yosys commit 151f7533 by @eddiehung, but that's just the commit introducing the testcase itself.

I believe the more stringent Fedora flags expose some sort of "off-by-one" or bounds checking issue somewhere in the code base. I'd be happy to help troubleshoot and/or test further if needed.

Thanks!!!

```
/usr/include/c++/8/bits/stl_vector.h:932: std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](std::vector<_Tp, _Alloc>::size_type) [with _Tp = Yosys::RTLIL::State; _Alloc = std::allocator<Yosys::RTLIL::State>; std::vector<_Tp, _Alloc>::reference = Yosys::RTLIL::State&; std::vector<_Tp, _Alloc>::size_type = long unsigned int]: Assertion '__builtin_expect(__n < this->size(), true)' failed.
make: *** [run-test.mk:29: run-bug1531.ys] Aborted (core dumped)
```

I can reproduce it.

--- a/frontends/ast/simplify.cc
+++ b/frontends/ast/simplify.cc
@@ -1200,16 +1200,22 @@ bool AstNode::simplify(bool const_fold, bool at_zero, bool in_lvalue, int stage,
 		if (varbuf->type != AST_CONSTANT)
 			log_file_error(filename, linenum, "Right hand side of 1st expression of generate for-loop is not constant!\n");
 
-		varbuf = new AstNode(AST_LOCALPARAM, varbuf);
-		varbuf->str = init_ast->children[0]->str;
-
 		auto resolved = current_scope.at(init_ast->children[0]->str);
 		if (resolved->range_valid) {
-			varbuf->range_left = resolved->range_left;
-			varbuf->range_right = resolved->range_right;
-			varbuf->range_swapped = resolved->range_swapped;
-			varbuf->range_valid = resolved->range_valid;
+			int const_size = varbuf->range_left - varbuf->range_right;
+			int resolved_size = resolved->range_left - resolved->range_right;
+			if (const_size < resolved_size) {
+				for (int i = const_size; i < resolved_size; i++)
+					varbuf->bits.push_back(resolved->is_signed ? varbuf->bits.back() : State::S0);
+				varbuf->range_left = resolved->range_left;
+				varbuf->range_right = resolved->range_right;
+				varbuf->range_swapped = resolved->range_swapped;
+				varbuf->range_valid = resolved->range_valid;
+			}
 		}
+
+		varbuf = new AstNode(AST_LOCALPARAM, varbuf);
+		varbuf->str = init_ast->children[0]->str;
 
 		AstNode *backup_scope_varbuf = current_scope[varbuf->str];
 		current_scope[varbuf->str] = varbuf;
@@ -3022,14 +3028,6 @@ void AstNode::expand_genblock(std::string index_var, std::string prefix, std::ma
 			current_ast_mod->children.push_back(p);
 			str = p->str;
 			id2ast = p;
-
-			auto resolved = current_scope.at(index_var);
-			if (resolved->range_valid) {
-				p->range_left = resolved->range_left;
-				p->range_right = resolved->range_right;
-				p->range_swapped = resolved->range_swapped;
-				p->range_valid = resolved->range_valid;
-			}
 		}
 	}
 

